agent/engine/dockerauth/ecr.go

```diff
@@ -17,54 +17,115 @@ import (
 	"encoding/base64"
 	"fmt"
 	"strings"
+	"sync"
+	"time"
 
 	"github.com/aws/amazon-ecs-agent/agent/api"
+	"github.com/aws/amazon-ecs-agent/agent/async"
 	"github.com/aws/amazon-ecs-agent/agent/ecr"
 	ecrapi "github.com/aws/amazon-ecs-agent/agent/ecr/model/ecr"
+	"github.com/aws/amazon-ecs-agent/agent/utils"
 	"github.com/aws/aws-sdk-go/aws"
 	log "github.com/cihub/seelog"
 	docker "github.com/fsouza/go-dockerclient"
 )
 
+type cacheKey struct {
+	region           string
+	roleARN          string
+	registryID       string
+	endpointOverride string
+}
+
 type ecrAuthProvider struct {
-	authData *api.ECRAuthData
-	client   ecr.ECRClient
+	tokenCache async.Cache
+	factory    ecr.ECRFactory
+	cacheLock  sync.RWMutex
```

Comment:
What's the cacheLock for?  Looking at `async.Cache`, it has its own lock internally.